### PR TITLE
[ceph-volume]: remove unneeded call to get_devices()

### PR DESCRIPTION
**Pull request by mgariepy**
ceph-volume lvm batch, there is no need to probe the device 
to generate the argparse help message.

Signed-off-by: Marc Gariepy <gariepy.marc@gmail.com>
Fixes: https://tracker.ceph.com/issues/47502

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by guits** [comment]
@andrewschoen @jan--f 

according to @mgariepy testing, it seems this patch can help to get around the issue seen in https://bugzilla.redhat.com/show_bug.cgi?id=1878500

Although I don't think this is the real fix here and I couldn't verify this myself, I would say as an immediate workaround it seems it can help to avoid this rotational detection issue.
By the way, what is the value of displaying the devices list in the helper? Isn't the role of subcommand `ceph-volume inventory`?

**Comment by andrewschoen** [comment]
> @andrewschoen @jan--f
> 
> according to @mgariepy testing, it seems this patch can help to get around the issue seen in https://bugzilla.redhat.com/show_bug.cgi?id=1878500
> 
> Although I don't think this is the real fix here and I couldn't verify this myself, I would say as an immediate workaround it seems it can help to avoid this rotational detection issue.
> By the way, what is the value of displaying the devices list in the helper? Isn't the role of subcommand `ceph-volume inventory`?

I don't think there is any value to that. Honestly, I think we should work to minimize the calls to `get_devices`. I've been working on this BZ https://bugzilla.redhat.com/show_bug.cgi?id=1822134 which I think is also related to subsequent calls to `get_devices` which results in many udev messages because of the opening and closing of raw devices in https://github.com/ceph/ceph/blob/master/src/ceph-volume/ceph_volume/util/disk.py#L693. In environments with numerous devices the udev messages seem to cause race conditions that can reset permissions on devices owned by ceph, as seen in BZ1822134. Those udev messages also seemed to be the cause of https://bugzilla.redhat.com/show_bug.cgi?id=1878500 which I've been working on with @guits.

I've been working on a patch (a6e030c8a260f50eff1293818335a502ded87046) that caches the return value of `get_devices` which should help, but I think we should also try to remove any unnecessary calls as well.

**Comment by jan--f** [comment]
> I've been working on a patch ([a6e030c](https://github.com/ceph/ceph/commit/a6e030c8a260f50eff1293818335a502ded87046)) that caches the return value of `get_devices` which should help, but I think we should also try to remove any unnecessary calls as well.

I just whipped up something similar, just py3 specific in 9c13dcc748364c0dd82d28d6af0990ad30cdf393. Issue is that this is actually slower then before (going by `ceph-volume inventory`), though the udev issue might be worked around by this (maybe @mgariepy  can confirm?).

I think caching `get_devices()` is the way to go, though something dead simple is probably enough.

File: src/ceph-volume/ceph_volume/devices/lvm/batch.py [hunk 1]
```diff
@@ -143,7 +143,7 @@ def __init__(self, argv):
         parser = argparse.ArgumentParser(
             prog='ceph-volume lvm batch',
             formatter_class=argparse.RawDescriptionHelpFormatter,
-            description=self.print_help(),
+            description="Utility for deploying osd in batch.",
```
Comment (jan--f) [inline src/ceph-volume/ceph_volume/devices/lvm/batch.py:146]: Please keep using the help text in `self._help`. Just drop the format and remove the placeholder at https://github.com/ceph/ceph/blob/b86b073b919ee22e80edec72d1ef6daed65dfa09/src/ceph-volume/ceph_volume/devices/lvm/batch.py#L131 plus the two surrounding lines.

**Review by jan--f — CHANGES_REQUESTED** [review]
Your commit has two `Signed-off-by` lines and please change the `Fixes:` line to `Fixes: https://tracker.ceph.com/issues/47502`.

**Comment by andrewschoen** [comment]
> > I've been working on a patch ([a6e030c](https://github.com/ceph/ceph/commit/a6e030c8a260f50eff1293818335a502ded87046)) that caches the return value of `get_devices` which should help, but I think we should also try to remove any unnecessary calls as well.
> 
> I just whipped up something similar, just py3 specific in [9c13dcc](https://github.com/ceph/ceph/commit/9c13dcc748364c0dd82d28d6af0990ad30cdf393). Issue is that this is actually slower then before (going by `ceph-volume inventory`), though the udev issue might be worked around by this (maybe @mgariepy can confirm?).
> 
> I think caching `get_devices()` is the way to go, though something dead simple is probably enough.

I'm fine with a simpler solution, I just thought the decorator could perhaps be useful for other things in the future.

I modified my commit with a better description (10d3f3f917fefbf8021621ff92fd92f7a23768d9). 

Limiting the amount of udev messages definitely seems to fix the issues @guits and I have been working on.

**Comment by andrewschoen** [comment]
@guits opened https://github.com/ceph/ceph/pull/37319 as a possible solution as well.

File: src/ceph-volume/ceph_volume/devices/lvm/batch.py (hunk 1 above)
Comment (mgariepy) [reply src/ceph-volume/ceph_volume/devices/lvm/batch.py]: done in the updated push.

**Review by guits — APPROVED** [review]

**Comment by guits** [comment]
jenkins test ceph-volume tox

**Comment by guits** [comment]
jenkins test ceph-volume all

**Comment by mgariepy** [comment]
jenkins test ceph-volume tox
jenkins test ceph-volume all

**Comment by guits** [comment]
jenkins test ceph-volume all

**Comment by guits** [comment]
jenkins test ceph-volume tox

**Comment by mgariepy** [comment]
jenkins test ceph-volume all

**Comment by mgariepy** [comment]
jenkins test ceph-volume tox

File: src/ceph-volume/ceph_volume/devices/lvm/batch.py [hunk 2]
```diff
@@ -260,17 +257,8 @@ def __init__(self, argv):
...
-        devices = [(device, details) for device, details in
-                       disk.get_devices().items() if details.get('partitions') == {}]
-        size_sort = lambda x: (x[0], x[1]['size'])
-        return device_formatter(sorted(devices, key=size_sort))
-
     def print_help(self):
-        return self._help.format(
-            detected_devices=self.get_devices(),
-        )
+        return self._help
```
Comment (jan--f) [inline src/ceph-volume/ceph_volume/devices/lvm/batch.py:261]: Let's get rid of `print_help()`.
Simply assign `description=self._help,` in the `ArgumentParser` constructor call..

**Review by jan--f — CHANGES_REQUESTED** [review]
Sorry, one last nit.

**Comment by mgariepy** [comment]
> Sorry, one last nit.

done, thanks for your review.

**Comment by mgariepy** [comment]
jenkins test ceph-volume tox

jenkins test ceph-volume all

**Comment by jan--f** [comment]
jenkins test ceph-volume tox

**Comment by jan--f** [comment]
jenkins test ceph-volume all

**Comment by guits** [comment]
jenkins test ceph-volume all

File: src/ceph-volume/ceph_volume/devices/lvm/batch.py [hunk 3]
```diff
@@ -260,18 +257,6 @@ def __init__(self, argv):
         for dev_list in ['', 'db_', 'wal_', 'journal_']:
             setattr(self, '{}usable'.format(dev_list), [])
 
-    def get_devices(self):
-        # remove devices with partitions
-        devices = [(device, details) for device, details in
-                       disk.get_devices().items() if details.get('partitions') == {}]
```
Comment (guits) [inline src/ceph-volume/ceph_volume/devices/lvm/batch.py:266]: you have to remove the import L6 to `ceph_volume.util.disk` since it's now unused

**Review by guits — CHANGES_REQUESTED** [review]

**Comment by jan--f** [comment]
jenkins test ceph-volume tox

**Comment by jan--f** [comment]
jenkins test ceph-volume all

**Review by jan--f — APPROVED** [review]

**Review by guits — APPROVED** [review]

**Pull request merged**